ext/random/randomizer.c

```diff
@@ -0,0 +1,289 @@
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Go Kudo <g-kudo@colopl.co.jp>                                |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+
+#include "ext/standard/php_array.h"
+#include "ext/standard/php_string.h"
+
+#include "ext/spl/spl_exceptions.h"
+#include "Zend/zend_exceptions.h"
+
+static inline void randomizer_common_init(php_random_randomizer *randomizer, zend_object *engine_object) {
+	if (engine_object->ce->type == ZEND_INTERNAL_CLASS) {
+		/* Internal classes always php_random_engine struct */
+		php_random_engine *engine = php_random_engine_from_obj(engine_object);
+		
+		/* Copy engine pointers */
+		randomizer->algo = engine->algo;
+		randomizer->status = engine->status;
+	} else {
+		/* Self allocation */
+		randomizer->status = php_random_status_alloc(&php_random_algo_user, false);
+		php_random_status_state_user *state = randomizer->status->state;
+		zend_string *mname;
+		zend_function *generate_method;
+
+		mname = zend_string_init("generate", sizeof("generate") - 1, 0);
+		generate_method = zend_hash_find_ptr(&engine_object->ce->function_table, mname);
+		zend_string_release(mname);
+
+		/* Create compatible state */
+		state->object = engine_object;
+		state->generate_method = generate_method;
+
+		/* Copy common pointers */
+		randomizer->algo = &php_random_algo_user;
+
+		/* Mark self-allocated for memory management */
+		randomizer->is_userland_algo = true;
+	}
+}
+
+/* {{{ Random\Randomizer::__construct() */
+PHP_METHOD(Random_Randomizer, __construct)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_object *engine_object = NULL;
+	zval zengine_object;
+	zend_string *pname;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(engine_object, random_ce_Random_Engine);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Create default RNG instance */
+	if (!engine_object) {
+		engine_object = random_ce_Random_Engine_Secure->create_object(random_ce_Random_Engine_Secure);
+
+		/* No need self-refcount */
+		GC_DELREF(engine_object);
+	}
+
+	ZVAL_OBJ(&zengine_object, engine_object);
+
+	/* Write property */
+	pname = zend_string_init("engine", sizeof("engine") - 1, 0);
+	zend_std_write_property(&randomizer->std, pname, &zengine_object, NULL);
+	zend_string_release(pname);
```

Comment:
I'm not sure, but I believe `zend_update_property()` would be a little cleaner here, because that one actually looks up the `->write_property()` function, instead of just assuming `zend_std_write_property()` is the correct one.

Reply:
This cannot be done because it is `readonly` property. I would like to know if there is a way to do this.

Reply:
That *should* work. I did the same in this commit: 83610987046d5a5ffea2777853d4a4f2d3313387

The following passes all tests for me:

```
zend_update_property(random_ce_Random_Randomizer, Z_OBJ_P(ZEND_THIS), "engine", strlen("engine"), &zengine_object);
```

Reply:
I tried it and it was fine.
I must have done something wrong... fixed: 7b8e1ee5e1e1822ecba6fd9dc4febbe4b0309890
Thanks!

Reply:
`zend_update_property()` should only fail, if called on a read-only property multiple times.